kotlin-analysis/src/main/kotlin/org/jetbrains/dokka/analysis/KotlinAnalysis.kt

```diff
@@ -34,20 +59,30 @@ class DokkaAnalysisConfiguration(
 )
 fun KotlinAnalysis(context: DokkaContext): KotlinAnalysis = KotlinAnalysis(context.configuration.sourceSets, context.logger)
 
-interface KotlinAnalysis : SourceSetDependent<EnvironmentAndFacade> {
+interface KotlinAnalysis : SourceSetDependent<EnvironmentAndFacade>, Closeable {
     override fun get(key: DokkaSourceSet): EnvironmentAndFacade
     operator fun get(sourceSetID: DokkaSourceSetID): EnvironmentAndFacade
 }
 
+/**
+ * It does not close [defaultKotlinAnalysis].
+ */
 internal class KotlinAnalysisImpl(
-    private val environments: SourceSetDependent<EnvironmentAndFacade>
+    private val environments: SourceSetDependent<EnvironmentAndFacade>,
+    private val defaultKotlinAnalysis: KotlinAnalysis? = null
 ) : KotlinAnalysis, SourceSetDependent<EnvironmentAndFacade> by environments {
```

Comment:
If I understand correctly, there can be two cases:
1. `KotlinAnalysisImpl` used as the main analysis for the project. Here `defaultKotlinAnalysis` would be `null`.
2. `KotlinAnalysisImpl` used for analysis of some samples, which requires more/different configuration of the classpath/source roots. Here `defaultKotlinAnalysis` will NOT be null and will have the value from the first case.

If my understanding is correct, I would propose to create separate implementations for them. 

To demonstrate:

```kotlin
internal class ProjectKotlinAnalysis(
    private val environments: SourceSetDependent<EnvironmentAndFacade>
) : KotlinAnalysis, SourceSetDependent<EnvironmentAndFacade> by environments {

    override fun get(key: DokkaSourceSet): EnvironmentAndFacade {
        return environments[key] ?: throw IllegalStateException("Missing EnvironmentAndFacade for sourceSet $key")
    }

    override fun get(sourceSetID: DokkaSourceSetID): EnvironmentAndFacade {
        return environments.entries.first { (sourceSet, _) -> sourceSet.sourceSetID == sourceSetID }.value
    }

    override fun close() {
        environments.values.forEach(EnvironmentAndFacade::close)
    }
}

/**
 * Samples analysis is used for analyzing samples bla bla bla
 * ...
 * Sample analysis is short-lived, so it is closed before [projectKotlinAnalysis].
 * ...
 * [projectKotlinAnalysis] needs to be closed separately.
 */
internal class SamplesKotlinAnalysis(
    private val sampleEnvironments: SourceSetDependent<EnvironmentAndFacade>,
    private val projectKotlinAnalysis: ProjectKotlinAnalysis
) : KotlinAnalysis, SourceSetDependent<EnvironmentAndFacade> by sampleEnvironments {
    
    override fun get(key: DokkaSourceSet): EnvironmentAndFacade {
        return sampleEnvironments[key] ?: projectKotlinAnalysis[key]
    }

    override fun get(sourceSetID: DokkaSourceSetID): EnvironmentAndFacade {
        return sampleEnvironments.entries.firstOrNull { (sourceSet, _) -> sourceSet.sourceSetID == sourceSetID }?.value
            ?: projectKotlinAnalysis[sourceSetID]
    }

    override fun close() {
        sampleEnvironments.values.forEach(EnvironmentAndFacade::close)
    }
}
```